pkg/datapath/iptables/iptables.go

```diff
@@ -695,26 +677,99 @@ func (m *IptablesManager) installStaticProxyRules() error {
 	return err
 }
 
+var tproxyMatch = regexp.MustCompile("CILIUM_PRE_mangle .*cilium: TPROXY")
+
+// copies old proxy rules
+func (m *IptablesManager) copyProxyRules(oldChain string, match string) {
+	copyRules := func(prog string, table string, re *regexp.Regexp, match, oldChain, newChain string) {
```

Comment:
I had a bit of trouble putting my finger on it, but these callback functions seemed a bit odd to me. I think the reason is that they are inherently untestable. Since it's a variable inside another function, we can't validate that it is performing the right actions. I realize that the outer function could theoretically be tested, but then we definitely bring in the iptables aspect to this, manage that kernel dependency during testing and so on. In an ideal world I feel like it'd be nice to have full-on unit tests that provide examples of what the iptables rules might look like (even if we had one giant static ruleset defined at the top), then we could run various functions like the scan/match piece of this function, and unit test to validate that this has the right behaviour.

I'm not going to push on this since this is a pretty targeted fix which is aimed for backporting, but this is clearly an aspect of technical debt we've picked up along the way. We're already having trouble encouraging people to get involved in reviewing code like this. Sure, part of that is just the nature of iptables and the difficulty understanding  the repercussions. But there's also just large testing gaps so even if someone wanted to pick up this code and play around with it, it's difficult to reason about or validate even the simplest of operations.

Reply:
introduced this local function to avoid repeating code; technically it is not a callback function as we do not return it. Will make this a separate package level function and add unit testing with some way of mocking up the `prog`.